### PR TITLE
fix(demo): custom bottom bar component

### PR DESCRIPTION
**Pull request by flochtililoch**
Fix double rendering of the custom bottom bar component by refactoring the following pieces:
- context: switch to using the reducer pattern, and expose 3 methods: `getElementProps`, `setElementProps`, `setElement`.
- core component: render any child element of the custom "non-render" element, not just the assumed single-child
- custom element: set the `registered` attribute on the element upon setting the props in the context, to avoid re-render loops when the screen holding the custom element re-renders. This does not prevent server side updates of the tab bar.
- update the demo app to factorize the markup for the tab bar as well as to show case the server updates of the tab bar.

Special thanks to @ehenighan for the thorough investigation and help solving #956.

https://github.com/user-attachments/assets/756d8585-d563-4a23-b845-3b42a843c4e7

## Summary by CodeRabbit

- **New Features**
  - Introduced a notification badge design for the bottom tab bar.
  - Added a new tab bar component to manage notifications within the second tab.

- **Enhancements**
  - Improved the BottomTabBar's context management with a reducer pattern for better state handling.
  - Streamlined the BottomTabBar component's logic and rendering process.
  - Enhanced tab navigation with a modular approach for dynamic rendering.

- **Bug Fixes**
  - Resolved rendering loops in the BottomTabBar component.

- **Documentation**
  - Updated context exports to prioritize the new custom hook for easier access.

### CODE REVIEW

File: demo/src/Components/BottomTabBar.tsx [hunk 1]
```diff
@@ -28,9 +28,12 @@ const BottomTabBar = (props: HvComponentProps) => {
       );
       return;
     }
-    ctx.setElementProps?.(navigator, props);
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [navigator, props]); // Exclude ctx from dependencies to avoid infinite loop
+    // To avoid rendering loops, we only set the element props once
+    if (props.element.getAttribute('registered') !== 'true') {
```
Comment (adamstep) [inline demo/src/Components/BottomTabBar.tsx:32]: What happens when the screen triggers a `reload` action and potentially gets a new bottom tab bar from the server? Is that handled correctly?

Comment (flochtililoch) [reply demo/src/Components/BottomTabBar.tsx:32]: Yes, it is handled correctly - the video I attached to the PR description showcases this.

**Review by adamstep — APPROVED** [review]

File: demo/src/Core/BottomTabBar/index.tsx [hunk 2]
```diff
@@ -13,43 +11,37 @@ import type { Props } from './types';
...
-  if (!props || !child) {
+  const onUpdateCustom = useCallback(
+    (href, action, currentElement, opts) => {
+      if (action === 'swap' && opts?.newElement) {
+        if (currentElement.parentNode) {
+          const newElement = currentElement.parentNode as Element;
+          newElement.replaceChild(opts.newElement, currentElement);
+          setElement?.(id, newElement);
+        } else {
+          console.warn('Parent node is null. Cannot replace child element.');
```
Comment (hgray-instawork) [inline demo/src/Core/BottomTabBar/index.tsx:29]: Better to use Logging.warn. See implementations in places like https://github.com/Instawork/hyperview/blob/231688b0b8416c74f2afa57b0edfffa2dfebfd33/src/behaviors/hv-unselect-all/index.ts

I found other console calls in the new tab bar components which aren't part of this PR.
```suggestion
          console.warn('Parent node is null. Cannot replace child element.');
```

**Review by hgray-instawork — APPROVED** [review]

File: demo/src/Core/BottomTabBar/index.tsx (hunk 2 above)
Comment (flochtililoch) [reply demo/src/Core/BottomTabBar/index.tsx:29]: I don't believe the demo app has the logger set up already?

Comment (hgray-instawork) [reply demo/src/Core/BottomTabBar/index.tsx:29]: Disregard. This is not internal to hyperview.

**Pull request merged**